### PR TITLE
fix: Fix deadlock when __sse_queue__ QueueHandler receives logs

### PR DESCRIPTION
**Pull request by JoanFM**
Linked to #936

### CODE REVIEW

File: tests/unit/flow/test_flow_with_sse.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from jina.flow import Flow
+
+
+def test_flow_with_sse():
+    f = Flow(logserver=True).\
+        add(uses='BaseExecutor', parallel=1).add(uses='BaseExecutor', parallel=2).build()
+    with f:
```
Comment (JoanFM) [inline tests/unit/flow/test_flow_with_sse.py:7]: this test is blocked forever without the changes

**Comment by hanxiao** [comment]
note that github action may restrict the access to `0.0.0.0:5000`, you may want to switch another port, and/or use `127.0.0.1/localhost`

**Comment by JoanFM** [comment]
> note that github action may restrict the access to `0.0.0.0:5000`, you may want to switch another port, and/or use `127.0.0.1/localhost`

Ritght now the problem in local is that by using `pytest-repeat` when repeating this test more than once there is still a `deadlock`,

**Comment by hanxiao** [comment]
i think it's Flask problem, terminating Flask programmatically is not straightforward

File: tests/unit/flow/test_flow_with_sse.py (hunk 1 above)
Comment (JoanFM) [reply tests/unit/flow/test_flow_with_sse.py:10]: Still does not solve, when using `pokedex-with-bit` with logserver there is still a deadlock

File: extra-requirements.txt [hunk 2]
```diff
@@ -56,5 +56,6 @@ pytest-subtests:            test
 pytest-xdist==1.34.0:       test
 pytest-timeout:             test
 pytest-cov:                 test
+pytest-repeat               test
```
Comment (hanxiao) [inline extra-requirements.txt:59]:
```suggestion
pytest-repeat:               test
```

**Comment by hanxiao** [comment]
btw in `flow/__init__.py` 474, `daemon` should be set to false to ensure the eventloop is ended without interruption
```
self._sse_logger = threading.Thread(name='sentinel-sse-logger',
                                                target=start_sse_logger, daemon=True,
                                                args=(self.args.logserver_config,
                                                      self.yaml_spec))
```

**Comment by JoanFM** [comment]
> btw in `flow/__init__.py` 474, `daemon` should be set to false to ensure the eventloop is ended without interruption
> 
> ```python
> self._sse_logger = threading.Thread(name='sentinel-sse-logger',
>                                                 target=start_sse_logger, daemon=True,
>                                                 args=(self.args.logserver_config,
>                                                       self.yaml_spec))
> ```

Still same error when using pokedex

File: jina/peapods/pea.py [hunk 3]
```diff
@@ -367,7 +370,7 @@ def post_init(self):
         """
         pass
 
-    def close(self) -> None:
+    def send_terminate_signal(self) -> None:
```
Comment (hanxiao) [inline jina/peapods/pea.py:373]: please keep `close()`, as it is the convention `of context_manager`. It also allows people to manually close a CM without using `with`.

Suggest:

```python
def __exit__(self, exc_type, exc_val, exc_tb) -> None:
       self.close()

def close(self) -> None:
        self.send_terminate_signal()
        self.is_shutdown.wait()
             clear_queue()
             self.join()
```

Comment (JoanFM) [reply jina/peapods/pea.py:373]: Done!

**Comment by hanxiao** [comment]
looks like CI is taking forever, something is not closed correctly

**Comment by JoanFM** [comment]
> note that github action may restrict the access to `0.0.0.0:5000`, you may want to switch another port, and/or use `127.0.0.1/localhost`

may the timeout in CI be related to this?

**Comment by JoanFM** [comment]
It is `containerPeas` we are not properly closing now

**Review by hanxiao — APPROVED** [review]

**Review by deepankarm — APPROVED** [review]

**Pull request merged**